doc/man/man1/rholo.1
.\" RCSid "$Id: rholo.1,v 1.5 2011/01/12 23:35:58 greg Exp $"
.TH RHOLO 1 1/14/99 RADIANCE
.SH NAME
rholo - generate/view a RADIANCE holodeck
.SH SYNOPSIS
.B rholo
[
.B "\-n npr"
][
.B "\-o dev"
][
.B \-w
][
.B \-i
][
.B \-f
|
.B \-r
]
.B hdkfile
[
.B "varfile | + | -"
[
.B "VAR\=value .."
]
]
.SH DESCRIPTION
.I Rholo
is a program for generating and viewing holodeck files.
Similar to
.I rvu(1),
.I rholo
can compute views interactively, but unlike
.I rvu,
it reuses any and all information that was previously computed in
this or earlier runs using the given holodeck file,
.I hdkfile.
.PP
The
.I \-n
option sets the number of
.I rtrace(1)
processes to start for the calculation.
It defaults to zero, which means that no new rays will be calculated.
In general, it is unwise to start more processes than there are
processors on the system.
On a multiprocessing system with 4 or more processors, a value one
less than the total should yield optimal interactive rates on a
lightly loaded system.
.PP
The \-o
option sets the output device to use for display.
Currently, there are at least two display drivers available,
.I x11
and
.I glx.
If no output device is specified, then
.I rholo
will start a global calculation of the holodeck, filling it
in as time goes by.
The quality of the final holodeck will depend on how long
.I rholo
runs before it is interrupted or runs out of file space or time,
according to the variable settings described in the control variable
section, below.
If no output device and no processes are specified,
.I rholo
creates an empty holodeck using the given
.I varfile,
if present.
.PP
The
.I \-i
option provides for reading from the standard input.
Without a display driver, the input should consist only of views,
which will be used to limit which parts of the holodeck are rendered
in a batch calculation.
With a display driver, most of the commands understood by the driver
can be issued either from the operating window or the standard input.
These commands are described together with their window equivalents in
the display driver section following the control variable section.
.PP
The
.I \-f
option permits the given holodeck to be clobbered.
Without this option, giving both the holodeck file and a variable file (or "-")
will result in an error message if the holodeck exists, since giving
both implies that a new holodeck is being created.
(When reusing an existing holodeck, the variable values are taken
from the holodeck header, though some may be overriden by giving a "+"
in place of the variable file.)\0
Also, attempts to clear the holodeck using the interactive
"clobber" command will be permitted only if the
.I \-f
option is given on the initial command line.
.PP
The
.I \-r
option tells
.I rholo
to open the holodeck file read-only, which is the default if there
are no ray calculation processes.
If one or more
.I rtrace
processes are started with the
.I \-n
option and the
.I \-r
option is given or the specified holodeck is not writable by the user,
then any additional rays computed during the session will be discarded
rather than saved to the holodeck file.
.PP
One or more holodeck section boundaries are defined along with other
parameters in the holodeck file or,
if the holodeck is being created, the
.I rholo
control variable file,
.I varfile.
These section boundaries define where you may move, or at least,
where you will be able to see, since they determine where computed
rays are stored.
Additional variable settings may be added or overridden on the
command line following
.I varfile.
If no
.I varfile
is needed, a holodeck may still be created by giving a "-" on the
command line in place of the variable file.
If you wish to override some of the variable settings in an existing
holodeck, use a "+", followed by the new settings on the command line.
Upper case variables specified more than once will result in
a warning message (unless the
.I \-w
option is present),
and the last value given will be the one used, unless it would conflict
with something in an existing holodeck that cannot be changed, such as
the section boundaries.
Changing section boundaries requires creating a new holodeck using
.I rholo
without a
.I \-n
or
.I \-o
option, then running
.I rhcopy(1)
to fill the new holodeck with the old holodeck's contents.
.PP
The
.I \-w
option turns off warnings about multiply and misassigned variables.
.PP
Rendering variable assignments appear one per line in
.I varfile.
The name of the variable is followed by an equals sign
('=') and its value(s).
The end of line may be escaped with a backslash ('\\'), though it is
not usually necessary.
Variables that should have only one value are given in upper case.
Variables that may have multiple values are given in lower case.
Variables may be abbreviated by their first three letters.
Comments in
.I varfile
start with a pound sign ('#') and proceed to the end of line.
.SH "CONTROL VARIABLES"
The control variables, their interpretations and default values
are given below.
.TP 10n
.BR OCTREE
The name of the octree file.
The default name is the same as
.I hdkfile
but with any suffix replaced by ".oct".
This variable may also be read from
.I rad(1)
if the "RIF" variable is set.
(See below.)\0
.TP
.BR RIF
This variable specifies a
.I rad
input file to use as a source of rendering options and other
variable settings.
If given,
.I rholo
will execute
.I rad
and get the rendering options to later pass to
.I rtrace.
Besides prepending the
.I render
variable,
.I rholo
will also extract default settings for the common "OCTREE" variable,
and the "EYESEP" variable.
Following the file name, overriding variable settings may be given,
which will be passed to
.I rad
on the command line.
Settings with spaces in them should be enclosed in quotes.
The execution of
.I rad
will also update the contents of the octree, if necessary.
There is no default value for this variable.
.TP
.BR EYESEP
The interocular spacing for stereo viewing.
I.e., the world distance between the pupils of the left and right eyes.
There is no default value for this variable.
.TP
.BR section
A section is a parallelepiped given by an origin and three axis
vectors (i.e., 12 floating point values in world coordinates).
The axis vectors define the three edges attached to the origin vertex,
and the other edges and vertices are determined by the parallel wall
constraint.
A holodeck section is a region in which the user may freely move about
to obtain a view of what is outside that region.
In object rendering mode, a section may instead contain a detailed
object to be viewed from the outside.
The grid dimensions for each axis may also be given by three additional
integer arguments following the three axes.
Otherwise, if the grid dimensions are left out or any are unspecified
or zero, the "GRID" variable will be used to determine them from the
section axis lengths.
(See below.)\0
There is no default value for this variable, and it is required.
If multiple values are given, they will be used for multiple rendering
sections, which may or may not be connected, but should generally not
overlap.
The starting view for interactive display will be the center
of the first section facing the positive X direction unless
"OBSTRUCTIONS" is set to True, when the view will be placed
outside the first section.
(See below for this variable's definition.)\0
The third axis of the first section is also
used as the default "view up" vector.
.TP
.BR geometry
This variable is used to associate geometry from an
octree file with one or more sections.
The specified octree will be used by certain drivers (e.g., the "ogl" driver)
to display simplified geometry using hardware lighting during motion.
If this variable is not set, such drivers will use the main octree file,
which contains all the scene geometry.
This can be slow if the scene is complex, so use simplified geometry
with portals (described below) or specify a non-existent file to turn
geometry rendering off.
If there is just one setting of this variable, it will be used for all
sections.
If there are multiple settings, they will correspond to multiple sections.
.TP
.BR portals
This variable is used to associate portal geometry with one or more
sections, as required for simplified geometry in some drivers (e.g., "ogl").
The portal geometry itself is given in one or more RADIANCE scene files or
quoted commands beginning with an exclamation mark ('!'),
and the input may or may not include material definitons.
(I.e., the surfaces may be modified by "void" if there are no materials.)
A portal is an imaginary surface that intervenes between a view and
some detailed geometry not included in the current section.
(See the "geometry" variable definition, above.)
Portals are often placed in doorways, windows and in front of mirrors.
Portal geometry may also be placed around local geometry that has been
culled due to its complexity.
This specification is necessary in order that the detail geometry be
drawn correctly, and that mirrors will work with virtual distances.
(See the definition of "VDISTANCE," below.)
The orientation of the portal surface geometry is ignored, so they
have effect no matter which way they are facing.
If there is just one setting of this variable, it will be used for all
sections.
If there are multiple settings, they will correspond to multiple sections.
.TP
.BR GRID
The default section grid size in world distance units.
If any section axis grid is unspecified, the length of the axis
will be divided by this number and rounded up to the next larger integer.
The grid size is a very important determiner of holodeck performance,
since the holodeck beam index is proportional to average axis grid dimension
to the fourth power!
If the beam index is too large, poor file and memory performance will
result.
If the beam index is too small, the holodeck resolution will suffer and
objects will tend to break up.
In general, the grid size should divide each section wall
into 64 or fewer cells for optimal performance.
The default value for this variable is the maximum section axis length
divided by 8.
.TP
.BR OBSTRUCTIONS
This boolean variable tells
.I rholo
whether or not to compute intersections with objects inside holodeck
sections.
If it is set to "False", then only objects outside the holodeck sections
will be visible.
This is appropriate when you know all sections to be devoid of geometry,
or when some secondary method is available for rendering geometry inside
each section.
If it is set to "True," all inside geometry will be visible.
There is no default for this variable, which means that rays will be
started at random points within each holodeck section, allowing interior
geometry to be partially sampled.
.TP
.BR VDISTANCE
This boolean variable determines whether the actual distance to objects
is computed, or the virtual distance.
If it is set to "True," the virtual distance will be used, which will
make reflections and refractions through smooth, flat objects clear,
but will blur the boundaries of those objects.
Note that some drivers cannot render virtual samples without the proper
placement of "portals" in the scene.
(See above for the definition of the "portals" variable.)
If it is set to "False," the reflections and refractions will be blurred,
but object boundaries will remain sharp.
The default value for this variable is "False."
.TP
.BR CACHE
The memory cache size to use for ray samples
during interactive rendering, in Megabytes.
This tuning parameter determines the tradeoff between memory use and
disk access time for interactive display.
This value will not affect memory use or performance for global
holodeck rendering if there is no display process.
The default cache is effectively set to 16 Megabytes.
If this variable is set to zero, no limit will be placed on memory
use and the process will grow to accommodate all the beams that
have been accessed.
.TP
.BR DISKSPACE
Specifies the maximum holodeck file size, in Megabytes.
Once the holodeck file reaches this size,
.I rtrace
will exit.
If there is no display process,
.I rholo
will also exit.
The default value for this variable is 0, which is interpreted as no
size limit.
.TP
.BR TIME
Sets the maximum time to run rtrace, in decimal hours.
After this length of time,
.I rtrace
will exit.
If there is no display process,
.I rholo
will also exit.
If there is a display process, and
.I rtrace
is restarted with the "restart" command, then the time clock will
be restarted as well.
The default value for this variable is 0, which is interpreted as no
time limit.
.TP
.BR REPORT
This variable may be used to specify a interval for
progress reports in minutes.
If this value is zero, then progress reports will not be given
in intervals, but a final report of the file size and fragmentation
will be issued when the program terminates, along with the number
of rays and packets computed.
If a filename is given after the interval, it will be used as the
error file for reports and error messages instead of the standard error.
There is no default value for this variable.
.TP
.BR render
This variable may be used to specify additional options to
.I rtrace.
These options will appear after the options set automatically by
.I rad,
and thus will override the default values.
.SH "DISPLAY DRIVER"
.I Rholo
may be started in interactive mode using the
.I \-o
option to specify an output display driver.
Currently, three drivers are supported on most machines,
.I glx,
.I ogl
and
.I x11.
(In addition, there are variations on the first two drivers for
stereo displays, local objects and human tone mapping.
These are accessed with some combination of the 's', 'o' and 'h' suffixes,
always in that order.
E.g., the OpenGL stereo driver with human tone mapping would be "oglsh".)
Each driver accepts simple one-character commands and mouse
view control in its operating window.
If the
.I \-i
option is also given, then the driver will also listen for commands entered
on the standard input.
(It is unwise to use the
.I \-i
option when rholo is run in the background, because it will occassionally
stop the process when input is available on the controlling terminal.)\0
The commands and their single-key window equivalents are given below.
.TP 10n
.BR "VIEW=    (mouse)"
Modify the current view with the specified parameters.
(See the
.I \-v*
view options in the
.I rpict(1)
manual page for parameter details.)\0
There is no one-character equivalent for this command in the display window.
Instead, the mouse is used to control the current view in the following ways:
.sp
.nf
CONTROL	MOUSE	ACTION
(none)	left	Move forward towards cursor position
(none)	middle	Rotate in place (usually safe)
(none)	right	Move backward away from cursor position
shift	left	Orbit left around cursor position
shift	middle	Orbit skyward
cntl	middle	Orbit earthward
shift	right	Orbit right around cursor position
cntl+shift	any		Frame focus by dragging rectangle
.fi
.sp
For all movements but rotating in place, the cursor must be placed over some
bit of visible geometry, otherwise the driver has no reference point from
which to work.
It is best to just experiment with these controls until you learn to fly
safely in your model.
And if you run into trouble, the "last" command is very useful.
(See below.)\0
.TP
.BR "last    'l'"
Return to the previous view.
Some drivers will save up multiple views in a history, but you are
guaranteed at least one.
.TP
.BR "where    'v'"
Print the current view parameters to the standard output.
This is useful for finding out where you are, or for saving specific
views in a keyframe file for animations or returning to later.
.TP
.BR "frame	'f'"
Change the calculation focus.
If the "frame" command is given with no arguments on the standard input,
it is equivalent to the interactive 'F' command, which releases
the current calculation focus.
If the "frame" command is followed by a relative horizontal and vertical
position (specified as floating point values between 0 and 1), then the
new focus is about this position on the screen (where 0 0 is at the lower
left of the display).
This is equivalent to the interactive 'f' command, which sets the focus
about the current window cursor position.
If four relative coordinates are given, they are assumed to mean the
minimum horizontal and vertical positon, and the maximum horizontal and
vertical position, in that order.
This is equivalent to dragging the mouse over a rectangular area
with the 'cntl+shift' keys held down.
.TP
.BR "pause    'p'"
Pause the ray calculation temporarily.
.TP
.BR "resume    <cr>"
Resume the ray calculation.
.TP
.BR "redraw    ^L"
Redraw the current view from values calculated and stored in the holodeck.
When executed from the display window via '^L', the effect may be slightly
different, since all stored information will be flushed.
.TP
.BR "kill    'K'"
Terminate the ray calculation process.
This is usually unnecessary, but is provided for special purpose applications.
.TP
.BR "restart    'R'"
Restart the ray calculation process.
If the "RIF" variable has been set,
.I rad
will be run first to assure that the octree is up to date.
.TP
.BR "clobber    'C'"
Clobber the holodeck contents, deleting all that has been calculated before.
To get an interactive dissolve of changes to the scene description,
use the sequence "kill," "clobber," "restart."
This command will be honored by
.I rholo
only if it was started with the
.I \-f
option.
.TP
.BR "quit    'q'"
Quit
.I rholo.
The ray tracing calculation is terminated and all values are flushed to
the holodeck file.
This is the normal way to exit the program.
.PP
In addition to these standard commands, all
drivers offer the following supplimentary controls.
.TP 10n
.BR "'h'"
Fix the head height.
All mouse-controlled view motions will be adjusted so that the head height
does not change (where vertical is determined by the current

src/hd/rholo.c
#ifndef lint
static const char	RCSid[] = "$Id: rholo.c,v 3.75 2010/09/30 15:43:30 greg Exp $";
#endif
/*
 * Radiance holodeck generation controller
 */

#include <stdio.h>
#include <time.h>
#include <signal.h>
#include <sys/stat.h>
#include <string.h>

#include "platform.h"
#include "rterror.h"
#include "resolu.h"
#include "rholo.h"
#include "random.h"

#ifndef FRAGWARN
#define FRAGWARN	20		/* fragmentation for warning (%) */
#endif
#ifndef MAXQTIME
#define MAXQTIME	3		/* target maximum seconds in queue */
#endif
					/* manual cache flushing frequency */
#ifndef RTFLUSH
#if MAXQTIME
#define RTFLUSH		(300/MAXQTIME*totqlen)	/* <= 5 minutes */
#else
#define RTFLUSH		(50*totqlen)		/* just guess */
#endif
#endif
			/* the following must be consistent with rholo.h */
int	NVARS = NRHVARS;		/* total number of variables */

VARIABLE	vv[] = RHVINIT;		/* variable-value pairs */

char	*progname;		/* our program name */
char	*hdkfile;		/* holodeck file name */
char	froot[256];		/* root file name */

int	ncprocs = 0;		/* desired number of compute processes */

char	*outdev = NULL;		/* output device name */

int	readinp = 0;		/* read commands from stdin */

int	force = 0;		/* allow overwrite of holodeck (-1 == read-only) */

time_t	starttime;		/* time we got started */
time_t	endtime;		/* time we should end by */
time_t	reporttime;		/* time for next report */

off_t	maxdisk;		/* maximum file space (bytes) */

int	rtargc = 1;		/* rtrace command */
char	*rtargv[128] = {"rtrace", NULL};

int	orig_mode = -1;		/* original file mode (-1 if unchanged) */

long	nraysdone = 0L;		/* number of rays done */
long	npacksdone = 0L;	/* number of packets done */

PACKET	*freepacks;		/* available packets */
int	totqlen;		/* maximum queue length (number of packets) */

char  *sigerr[NSIG];		/* signal error messages */

extern int	nowarn;		/* turn warnings off? */

static void onsig(int  signo);
static void sigdie(int  signo, char  *msg);
static int resfmode(int	fd, int	mod);
static void initrholo(void);
static int rholo(void);
static void setdefaults(HDGRID	*gp);
static void creatholo(HDGRID	*gp);
static gethfunc headline;
static void loadholo(void);
static void rootname(char	*rn, char	*fn);
static void badvalue(int	vc);


int
main(
	int	argc,
	char	*argv[]
)
{
	int	i;

	progname = argv[0];			/* get arguments */
	for (i = 1; i < argc && argv[i][0] == '-'; i++)
		switch (argv[i][1]) {
		case 'w':			/* turn off warnings */
			nowarn++;
			break;
		case 'f':			/* force overwrite */
			force = 1;
			break;
		case 'r':			/* read-only mode */
			force = -1;
			break;
		case 'i':			/* read input from stdin */
			readinp++;
			break;
		case 'n':			/* compute processes */
			if (i >= argc-2)
				goto userr;
			ncprocs = atoi(argv[++i]);
			break;
		case 'o':			/* output display */
			if (i >= argc-2)
				goto userr;
			outdev = argv[++i];
			break;
		default:
			goto userr;
		}
						/* get root file name */
	if (i >= argc)
		goto userr;
	rootname(froot, hdkfile=argv[i++]);
						/* load variables? */
	if (i < argc)
		if (argv[i][0] != '-' && argv[i][0] != '+')
			loadvars(argv[i]);	/* load variables from file */

	if (i >= argc || argv[i][0] == '+')
		loadholo();			/* load existing holodeck */

	while (++i < argc)			/* get command line settings */
		if (setvariable(argv[i], matchvar) < 0) {
			sprintf(errmsg, "unknown variable: %s", argv[i]);
			error(USER, errmsg);
		}
						/* check settings */
	checkvalues();
						/* load rad input file */
	getradfile();

	if (hdlist[0] == NULL) {		/* create new holodeck */
		HDGRID	hdg[HDMAX];
							/* set defaults */
		setdefaults(hdg);
							/* check read-only */
		if (force < 0)
			error(USER, "cannot create read-only holodeck");
							/* holodeck exists? */
		if (!force && access(hdkfile, R_OK|W_OK) == 0)
			error(USER,
				"holodeck file exists -- use -f to overwrite");
							/* create holodeck */
		creatholo(hdg);
	} else					/* else just set defaults */
		setdefaults(NULL);
						/* initialize */
	initrholo();
						/* main loop */
	while (rholo())
		;
						/* done */
	quit(0);
userr:
	fprintf(stderr,
"Usage: %s [-n nprocs][-o disp][-i][-w][-r|-f] output.hdk [control.hif|+|- [VAR=val ..]]\n",
			progname);
	quit(1);
	return 1; /* pro forma return */
}


static void
onsig(				/* fatal signal */
	int  signo
)
{
	static int  gotsig = 0;

	if (gotsig > 1)			/* we're going as fast as we can! */
		return;
	if (gotsig++) {			/* two signals and we split */
		hdsync(NULL, 0);	/* don't leave w/o saying goodbye */
		_exit(signo);
	}
	alarm(300);			/* allow 5 minutes to clean up */
	eputs("signal - ");
	eputs(sigerr[signo]);
	eputs("\n");
	quit(3);
}


static void
sigdie(			/* set fatal signal */
	int  signo,
	char  *msg
)
{
	if (signal(signo, onsig) == SIG_IGN)
		signal(signo, SIG_IGN);
	sigerr[signo] = msg;
}


static int
resfmode(		/* restrict open file access mode */
	int	fd,
	int	mod
)
{
	struct stat	stbuf;
					/* get original mode */
	if (fstat(fd, &stbuf) < 0)
		error(SYSTEM, "cannot stat open holodeck file");
	mod &= stbuf.st_mode;		/* always more restrictive */
	if (mod == (stbuf.st_mode & 0777))
		return(-1);		/* already set */
					/* else change it */
	if (fchmod(fd, mod) < 0) {
		error(WARNING, "cannot change holodeck file access mode");
		return(-1);
	}
	return(stbuf.st_mode);		/* return original mode */
}


static void
initrholo(void)			/* get our holodeck running */
{
	extern int	global_packet();
	register int	i;
						/* close holodeck on exec() */
	fcntl(hdlist[0]->fd, F_SETFD, FD_CLOEXEC);

	if (outdev != NULL)			/* open output device */
		disp_open(outdev);
	else if (ncprocs > 0)			/* else use global ray feed */
		init_global();
						/* record disk space limit */
	if (!vdef(DISKSPACE))
		maxdisk = ((off_t)1<<(sizeof(off_t)*8-2)) - 1024;
	else
		maxdisk = 1024.*1024.*vflt(DISKSPACE);
						/* set up memory cache */
	if (outdev == NULL)
		hdcachesize = 0;		/* manual flushing */
	else if (vdef(CACHE))
		hdcachesize = 1024.*1024.*vflt(CACHE);
						/* open report file */
	if (vdef(REPORT)) {
		register char	*s = sskip2(vval(REPORT), 1);
		if (*s && freopen(s, "a", stderr) == NULL)
			quit(2);
	}
						/* mark the starting time */
	starttime = time(NULL);
						/* compute end time */
	if (!vdef(TIME) || vflt(TIME) <= FTINY)
		endtime = 0;
	else
		endtime = starttime + vflt(TIME)*3600. + .5;
						/* start rtrace */
	if (ncprocs > 0) {
		totqlen = i = start_rtrace();
		if (i < 1)
			error(USER, "cannot start rtrace process(es)");
		if (vdef(REPORT)) {		/* make first report */
			printargs(rtargc, rtargv, stderr);
			report(0);
		}
						/* allocate packets */
		freepacks = (PACKET *)bmalloc(i*sizeof(PACKET));
		if (freepacks == NULL)
			goto memerr;
		freepacks[--i].nr = 0;
		freepacks[i].next = NULL;
		if (!vdef(OBSTRUCTIONS) || !vbool(OBSTRUCTIONS)) {
			freepacks[i].offset = (float *)bmalloc(
					RPACKSIZ*sizeof(float)*(i+1) );
			if (freepacks[i].offset == NULL)
				goto memerr;
		} else
			freepacks[i].offset = NULL;
		while (i--) {
			freepacks[i].nr = 0;
			freepacks[i].offset = freepacks[i+1].offset == NULL ?
					NULL : freepacks[i+1].offset+RPACKSIZ ;
			freepacks[i].next = &freepacks[i+1];
		}
	}
					/* set up signal handling */
	sigdie(SIGINT, "Interrupt");
	sigdie(SIGTERM, "Terminate");
#ifdef SIGHUP
	sigdie(SIGHUP, "Hangup");
#endif
#ifdef SIGPIPE
	sigdie(SIGPIPE, "Broken pipe");
#endif
#ifdef SIGALRM
	sigdie(SIGALRM, "Alarm clock");
#endif
#ifdef	SIGXCPU
	sigdie(SIGXCPU, "CPU limit exceeded");
#endif
#ifdef	SIGXFSZ
	sigdie(SIGXFSZ, "File size exceeded");
#endif
					/* protect holodeck file */
	orig_mode = resfmode(hdlist[0]->fd, (ncprocs>0) & (force>=0) ? 0 : 0444);
	return;
memerr:
	error(SYSTEM, "out of memory in initrholo");
}


static int
rholo(void)				/* holodeck main loop */
{
	static off_t	nextfragwarn = 100L<<20;
	static int	idle = 0;
	PACKET	*pl = NULL, *plend;
	off_t	fsiz;
	int	pksiz;
	register PACKET	*p;
	time_t	t;
					/* check display */
	if (nprocs <= 0)
		idle = 1;
	if (outdev != NULL) {
		if (!disp_check(idle))
			return(0);	/* quit request */
		if (nprocs <= 0)
			return(1);
	} else if (idle)
		return(0);		/* all done */
	fsiz = hdfilen(hdlist[0]->fd);	/* check file size */
	if (maxdisk > 0 && fsiz >= maxdisk) {
		error(USER, "file limit exceeded");
		done_rtrace();
		return(1);	/* comes back */
	}
#if FRAGWARN
	if (fsiz >= nextfragwarn) {
		double	pctfrag = 100.*(fsiz-hdfiluse(hdlist[0]->fd))/fsiz;
	       	if (pctfrag >= (double)FRAGWARN) {
			sprintf(errmsg, "holodeck file fragmentation is %.0f%%",
					pctfrag);
			error(WARNING, errmsg);
			nextfragwarn = fsiz + (fsiz>>2);
		} else
			nextfragwarn = fsiz + (10L<<20);
	}
#endif
	t = time(NULL);			/* check time */
	if (endtime > 0 && t >= endtime) {
		error(USER, "time limit exceeded");
		done_rtrace();
		return(1);	/* comes back */
	}
	if (reporttime > 0 && t >= reporttime)
		report(t);
					/* figure out good packet size */
	pksiz = RPACKSIZ;
#if MAXQTIME
	if (!chunkycmp) {
		pksiz = nraysdone*MAXQTIME/(totqlen*(t - starttime + 1L));
		if (pksiz < 1) pksiz = 1;
		else if (pksiz > RPACKSIZ) pksiz = RPACKSIZ;
	}
#endif
	idle = 0;			/* get packets to process */
	while (freepacks != NULL) {
		p = freepacks; freepacks = p->next; p->next = NULL;
		if (!next_packet(p, pksiz)) {
			p->next = freepacks; freepacks = p;
			idle = 1;
			break;
		}
		if (pl == NULL) pl = p;
		else plend->next = p;
		plend = p;
	}
					/* process packets */
	done_packets(do_packets(pl));
	return(1);			/* and continue */
}


static void
setdefaults(			/* set default values */
	register HDGRID	*gp
)
{
	extern char	*atos();
	register int	i;
	int	n;
	double	len[3], d;

	if (!vdef(SECTION)) {
		sprintf(errmsg, "%s must be defined", vnam(SECTION));
		error(USER, errmsg);
	}
	if (!vdef(OCTREE)) {
		if ((vval(OCTREE) = bmalloc(strlen(froot)+5)) == NULL)
			error(SYSTEM, "out of memory");
		sprintf(vval(OCTREE), "%s.oct", froot);
		vdef(OCTREE)++;
	}
	if (!vdef(VDIST)) {
		vval(VDIST) = "F";
		vdef(VDIST)++;
	}
				/* append rendering options */
	if (vdef(RENDER))
		rtargc += wordstring(rtargv+rtargc, vval(RENDER));
	
	if (gp == NULL)		/* already initialized? */
		return;
				/* set grid parameters */
	for (n = 0; n < vdef(SECTION); n++, gp++) {
		gp->grid[0] = gp->grid[1] = gp->grid[2] = 0;
		if (sscanf(nvalue(SECTION, n),
		"%lf %lf %lf %lf %lf %lf %lf %lf %lf %lf %lf %lf %hd %hd %hd",
				&gp->orig[0], &gp->orig[1], &gp->orig[2],
				&gp->xv[0][0], &gp->xv[0][1], &gp->xv[0][2],
				&gp->xv[1][0], &gp->xv[1][1], &gp->xv[1][2],
				&gp->xv[2][0], &gp->xv[2][1], &gp->xv[2][2],
				&gp->grid[0], &gp->grid[1], &gp->grid[2]) < 12)
			badvalue(SECTION);
		for (i = 0; i < 3; i++)
			len[i] = VLEN(gp->xv[i]);
		if (!vdef(GRID)) {
			d = 2/2e5*( len[0]*len[0]*(len[1]*len[1] +
					len[2]*len[2] + 4*len[1]*len[2])
				+ len[1]*len[1]*len[2]*(len[2] + 4*len[0])
				+ 4*len[0]*len[1]*len[2]*len[2] );
			d = sqrt(sqrt(d));
		} else if ((d = vflt(GRID)) <= FTINY)
			badvalue(GRID);
		for (i = 0; i < 3; i++)
			if (gp->grid[i] <= 0)
				gp->grid[i] = len[i]/d + (1.-FTINY);
	}
}


static void
creatholo(			/* create a holodeck output file */
	HDGRID	*gp
)
{
	extern char	VersionID[];
	off_t	lastloc, nextloc;
	int	n;
	int	fd;
	FILE	*fp;
					/* open & truncate file */
	if ((fp = fopen(hdkfile, "w+")) == NULL) {
		sprintf(errmsg, "cannot open \"%s\" for writing", hdkfile);
		error(SYSTEM, errmsg);
	}
					/* write information header */
	newheader("RADIANCE", fp);
	fprintf(fp, "SOFTWARE= %s\n", VersionID);
	printvars(fp);
	fputformat(HOLOFMT, fp);
	fputc('\n', fp);
	putw(HOLOMAGIC, fp);		/* put magic number */
	fd = dup(fileno(fp));
	fclose(fp);			/* flush and close stdio stream */
	lastloc = lseek(fd, (off_t)0, SEEK_END);
	for (n = vdef(SECTION); n--; gp++) {	/* initialize each section */
		nextloc = 0L;
		write(fd, (char *)&nextloc, sizeof(nextloc));
		hdinit(fd, gp);			/* writes beam index */
		if (!n)
			break;
		nextloc = hdfilen(fd);		/* write section pointer */
		if (lseek(fd, lastloc, SEEK_SET) < 0)
			error(SYSTEM,
				"cannot seek on holodeck file in creatholo");
		write(fd, (char *)&nextloc, sizeof(nextloc));
		lseek(fd, (lastloc=nextloc), SEEK_SET);
	}
}


static int
headline(			/* process information header line */
	char	*s,
	void	*p
)
{
	extern char	FMTSTR[];
	register char	*cp;
	char	fmt[32];

	if (formatval(fmt, s)) {
		if (strcmp(fmt, HOLOFMT)) {
			sprintf(errmsg, "%s file \"%s\" has %s%s",
					HOLOFMT, hdkfile, FMTSTR, fmt);
			error(USER, errmsg);
		}
		return(0);
	}
	for (cp = s; *cp; cp++)		/* take off any comments */
		if (*cp == '#') {
			*cp = '\0';
			break;
		}
	setvariable(s, matchvar);	/* don't flag errors */
	return(0);
}


static void
loadholo(void)			/* start loading a holodeck from fname */
{
	FILE	*fp;
	int	fd;
	int	n;
	off_t	nextloc;
	
	if ((ncprocs > 0) & (force >= 0))
		fp = fopen(hdkfile, "r+");
	else
		fp = NULL;
	if (fp == NULL) {
		if ((fp = fopen(hdkfile, "r")) == NULL) {
			sprintf(errmsg, "cannot open \"%s\"", hdkfile);
			error(SYSTEM, errmsg);
		}
		if (ncprocs > 0) {
			sprintf(errmsg, "\"%s\" is read-only", hdkfile);
			if (outdev == NULL)
				error(USER, errmsg);
			strcat(errmsg, "; new rays will be discarded");
			error(WARNING, errmsg);
			force = -1;
		}
	}
					/* load variables from header */
	getheader(fp, headline, NULL);
					/* check magic number */
	if (getw(fp) != HOLOMAGIC) {
		sprintf(errmsg, "bad magic number in holodeck file \"%s\"",
				hdkfile);
		error(USER, errmsg);
	}
	nextloc = ftell(fp);			/* get stdio position */
	fd = dup(fileno(fp));
	fclose(fp);				/* done with stdio */
	for (n = 0; nextloc > 0L; n++) {	/* initialize each section */
		lseek(fd, nextloc, SEEK_SET);
		read(fd, (char *)&nextloc, sizeof(nextloc));
		hdinit(fd, NULL);
	}
	if (n != vdef(SECTION)) {
		sprintf(errmsg, "number of sections does not match %s setting",
				vnam(SECTION));
		error(WARNING, errmsg);
	}
}


extern void
done_packets(		/* handle finished packets */
	PACKET	*pl
)
{
	static int	n2flush = 0;
	register PACKET	*p;

	while (pl != NULL) {
		p = pl; pl = p->next; p->next = NULL;
		if (p->nr > 0) {		/* add to holodeck */
			memcpy( (void *)hdnewrays(hdlist[p->hd],p->bi,p->nr),
				(void *)p->ra,
				p->nr*sizeof(RAYVAL));
			if (outdev != NULL)	/* display it */
				disp_packet((PACKHEAD *)p);
			if (hdcachesize <= 0)
				n2flush++;
			nraysdone += p->nr;
			npacksdone++;
			p->nr = 0;
		}
		p->next = freepacks;		/* push onto free list */
		freepacks = p;
	}
	if (n2flush >= RTFLUSH) {
		if (outdev != NULL)
			hdsync(NULL, 1);
		else
			hdflush(NULL);
		n2flush = 0;
	}
}


static void
rootname(		/* remove tail from end of fn */
	register char	*rn,
	register char	*fn
)
{
	char	*tp, *dp;

	for (tp = NULL, dp = rn; (*rn = *fn++); rn++) {
		if (*rn == '/')
			dp = rn;
		else if (*rn == '.')
			tp = rn;
	}
	if (tp != NULL && tp > dp)
		*tp = '\0';
}


static void
badvalue(			/* report bad variable value and exit */
	int	vc
)
{
	sprintf(errmsg, "bad value for variable '%s'", vnam(vc));
	error(USER, errmsg);
}


void
eputs(s)			/* put error message to stderr */
register char  *s;
{
	static int  midline = 0;

	if (!*s)
		return;
	if (!midline++) {	/* prepend line with program name */
		fputs(progname, stderr);
		fputs(": ", stderr);
	}
	fputs(s, stderr);
	if (s[strlen(s)-1] == '\n') {
		fflush(stderr);
		midline = 0;
	}
}


void
quit(ec)			/* exit program gracefully */
int	ec;
{
	int	status = 0;

	if (hdlist[0] != NULL) {	/* close holodeck */
		if (nprocs > 0)
			status = done_rtrace();		/* calls hdsync() */
		if ((ncprocs > 0) & (force >= 0) && vdef(REPORT)) {
			off_t	fsiz, fuse;
			fsiz = hdfilen(hdlist[0]->fd);
			fuse = hdfiluse(hdlist[0]->fd);
			fprintf(stderr,
			"%s: %.1f Mbyte holodeck file, %.1f%% fragmentation\n",
					hdkfile, fsiz/(1024.*1024.),
					100.*(fsiz-fuse)/fsiz);
		}
	}
	if (orig_mode >= 0)		/* reset holodeck access mode */
		fchmod(hdlist[0]->fd, orig_mode);
	if (outdev != NULL)		/* close display */
		disp_close();
	exit(ec ? ec : status);		/* exit */
}
